### PR TITLE
Fix wrong swappiness value from REST api

### PR DESCRIPTION
**Pull request by hqhq**
We added memory swappiness support recently, but for old
usage without MemorySwappiness field in json, the default
would be 0, and assigned to cgroup, which is not excepted.

We need to add the real default value -1 before decode
json file.

Signed-off-by: Qiang Huang h.huangqiang@huawei.com

### CODE REVIEW

**Comment by hqhq** [comment]
I'm not sure about this PR, the fix breaks backward compatibility for HostConfig, before this, if user didn't use `HostConfig` in json file, we will take all HostConfig entries in the json file to `HostConfig`, it is done by `decoder.Decode()`. After this fix, decode result always has `HostConfig` entry, so we need to deal with that by hand.

I feel pretty bad about the second commit, it hasn't handled all `HostConfig` entries, I'm seeking for other options or see if we are ok with this approach.

**Comment by estesp** [comment]
I think the intent is absolutely correct--old HostConfig json contents will not have a value, and the default == 0 is definitely wrong and will produce unexpected behavior.

The open question, given that this PR shows some of the tricky issues around handling that, is whether this means we should be looking at changing how `memory.swappiness` is represented so a "default value" is not always assumed to mean "is set"?

**Comment by hqhq** [comment]
@estesp Yes we've been facing this problem for a long time, nothing really hurt until swappiness appears. And I know there'll be more problems if we support dynamically change container configs using current approach. Because some bool configs have false by default, if they are enabled in runtime and we send a whole struct which might contain the default false to change configs. We'll get unexpected result.

@duglin used to have an idea that we always send user requests as json, not as a whole struct, and daemon parse it one field by another, not just decode it to a struct. It might be a bit complicate but I think it'll fix this kind of problems, because we'll know what options users really send and what values they want them be, no more defaults.

**Comment by duglin** [comment]
@hqhq yep - when you send an entire struct you'll eventually get into the situation where you can't tell if the value there is some kind of default value, or if it means "not set", or if that's the value the send actually specified.

**Comment by LK4D4** [comment]
makes sense for me
ping @docker/core-maintainers to move to code-review

**Comment by estesp** [comment]
Yeah--so this PR can't go forward until we figure out how to handle that 2nd commit properly.. right now, for example, this test is broken: https://github.com/docker/docker/blob/master/integration-cli/docker_api_containers_test.go#L1034-L1039  because `CpusetCpus` will be "overwritten" by empty `InnerHostConfig` as it doesn't have a check added to the list that @hqhq added already.

I don't have enough history of `HostConfig` and the backward compat. requirements to know whether this is the proper approach or not, but definitely if we do it this way, we have to handle all possible values that will be filled in `HostConfig` because of the JSON decoder.

**Comment by estesp** [comment]
@calavera I think 1.8 needs to deal with this in some way or else we would need to deprecate certain (currently supported) distros from being able to use Docker with pre-existing containers (e.g. upgrade to 1.8 and restart containers will fail).  Adding to the 1.8 milestone to get the discussion going, but of course if agreement is it can't be fixed for this release, we need some docs/notice on what won't work.

**Comment by calavera** [comment]
@estesp thanks. I agree that this should be in 1.8.

**Comment by crosbymichael** [comment]
Why do we need the second commit?

**Comment by estesp** [comment]
@crosbymichael I don't have a lot of the backstory of `HostConfig` values being "outside" of a `HostConfig: {}` object in the JSON (apparently that's deprecated but still needs to be supported?), but for sure this test (*docker_api_containers_test.go:1072: DockerSuite.TestContainerApiCreateWithCpuSharesCpuset*) expects the value in the bare JSON to be reflected in the final `HostConfig` object returned by `getHostConfig()`, and the only way to do that after the initial change is to copy value by value from `hc` (which are any values in the decoded JSON which match the name of `HostConfig` struct members) into the **InnerHostConfig** which is wrapped in an actual `HostConfig: { }` section, from what I understand.

Maybe there is a cleaner way to do all of this, because right now it sort of stinks to have a default (non-existent) value represent a settable value.  E.g. `memory.swappiness == 0` has an actual meaning, so picking "-1" to mean unset works in general, but is problematic in the case of parsing a JSON stream in which it did not exist, and therefore defaults to `0`, which is the reason for the first commit.

**Comment by hqhq** [comment]
Rebased and added compatibility for CpusetCpus and VolumesFrom.

@crosbymichael As @estesp said, the second commit is for backward compatibility, some users still use entries belong to HostConfig without HostConfig entry, after the first commit, these usage will be all broken. The problem  with the second commit is how many entries do we need to handle.

**Comment by calavera** [comment]
@crosbymichael can you take a look at this?

File: runconfig/config.go [hunk 1]
```diff
@@ -204,6 +244,12 @@ func DecodeContainerConfig(src io.Reader) (*Config, *HostConfig, error) {
 	decoder := json.NewDecoder(src)
 
 	var w ContainerConfigWrapper
+
+	// Decode will assign type initial values to these absent fields, but for
+	// some json fields, type initial values have some meaning, we need to assign
+	// real default values before Decode them.
+	w.InnerHostConfig = &HostConfig{}
+	w.InnerHostConfig.MemorySwappiness = -1
```
Comment (icecrime) [inline runconfig/config.go]: Isn't this better hidden in a `func NewHostConfig() *HostConfig`?

**Comment by icecrime** [comment]
I don't understand the need for the second commit neither, sorry :-(

**Comment by icecrime** [comment]
I don't entirely understand this patch, sorry. We need to fix this with 1.8.0 so if others can confirm it does fix the issue, we should go for it, but it seems we need a better long term solution for API "nullable values".

**Comment by crosbymichael** [comment]
I don't think the second commit makes sense.  you cannot have the host config and inner host config at the same time.  If you pass the inner host config it nullifies everything else because it can be null in the json request, so we know that the user passed it.  

The fix should be to just make sure the swappieness defaults to -1 and it is only support on the new host config, there is no backward compat for swappiness because it's new.  It's either that or remove the swappiness support in this release till we can fix this stuff because the host config /inner host config is super confusing and badly designed.

**Comment by calavera** [comment]
what if we instead of doing this, we use a proper type for `MemorySwapiness` and use a pointer to it in the `HostConfig` struct? That way, if the json field is missing, the pointer will be nil and we won't asign the value to the cgroup.

**Comment by calavera** [comment]
please, take a look at https://github.com/docker/docker/pull/15134.

**Comment by icecrime** [comment]
Sorry @hqhq, closing in favor of the simpler #15134.

**Pull request closed without merging**